src/JsonApi/Serializer/ItemNormalizer.php

```diff
@@ -0,0 +1,561 @@
+            }
+
+            $context['object_to_populate'] = $this->iriConverter->getItemFromIri(
+                $data['data']['id'],
+                $context + ['fetch_data' => false]
+            );
+        }
+
+        // Merge attributes and relations previous to apply parents denormalizing
+        $dataToDenormalize = array_merge(
+            $data['data']['attributes'] ?? [],
+            $data['data']['relationships'] ?? []
+        );
+
+        return parent::denormalize(
+            $dataToDenormalize,
+            $class,
+            $format,
+            $context
+        );
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    protected function getAttributes($object, $format, array $context)
+    {
+        return $this->getComponents($object, $format, $context)['attributes'];
+    }
+
+    /**
+     * Gets JSON API components of the resource: attributes, relationships, meta and links.
+     *
+     * @param object      $object
+     * @param string|null $format
+     * @param array       $context
+     *
+     * @return array
+     */
+    private function getComponents($object, string $format = null, array $context)
+    {
+        if (isset($this->componentsCache[$context['cache_key']])) {
+            return $this->componentsCache[$context['cache_key']];
+        }
+
+        $attributes = parent::getAttributes($object, $format, $context);
+
+        $options = $this->getFactoryOptions($context);
+
+        $typeShortName = $className = '';
+
+        $components = [
+            'links' => [],
+            'relationships' => [],
+            'attributes' => [],
+            'meta' => [],
+        ];
+
+        foreach ($attributes as $attribute) {
+            $propertyMetadata = $this
+                ->propertyMetadataFactory
+                ->create($context['resource_class'], $attribute, $options);
+
+            $type = $propertyMetadata->getType();
+            $isOne = $isMany = false;
+
+            if (null !== $type) {
+                if ($type->isCollection()) {
+                    $valueType = $type->getCollectionValueType();
+
+                    $isMany = null !== $valueType
+                        && ($className = $valueType->getClassName())
+                        && $this->resourceClassResolver->isResourceClass($className);
+                } else {
+                    $className = $type->getClassName();
+
+                    $isOne = null !== $className
+                        && $this->resourceClassResolver->isResourceClass($className);
+                }
+
+                $typeShortName =
+                    (
+                        (
+                            null !== $className
+                                && $this->resourceClassResolver->isResourceClass($className)
+                        )
+                            ? $this->resourceMetadataFactory->create($className)->getShortName() :
+                            ''
+                    );
```

Comment:
IMO this is not readable enough, WDYT about using an `if` condition instead of ternary?